File: nodes/core/market/filters/adx_filter_node.py
from core.types_registry import IndicatorResult, IndicatorType, IndicatorValue, OHLCVBar
from nodes.core.market.filters.base.base_indicator_filter_node import BaseIndicatorFilter
from services.indicator_calculators.adx_calculator import calculate_adx


class ADXFilter(BaseIndicatorFilter):
    """
    Filters assets based on ADX (Average Directional Index) values.
    """

    default_params = {
        "min_adx": 25.0,
        "timeperiod": 14,
    }

    params_meta = [
        {
            "name": "min_adx",
            "type": "number",
            "default": 25.0,
            "min": 0.0,
            "max": 100.0,
            "step": 0.1,
        },
        {"name": "timeperiod", "type": "number", "default": 14, "min": 1, "step": 1},
    ]

    def _validate_indicator_params(self):
        min_adx = self.params.get("min_adx", 25.0)
        timeperiod = self.params.get("timeperiod", 14)
        if not isinstance(min_adx, (int, float)) or min_adx < 0:
            raise ValueError("Minimum ADX cannot be negative")
        if not isinstance(timeperiod, (int, float)) or timeperiod <= 0:
            raise ValueError("Time period must be positive")

    def _calculate_indicator(self, ohlcv_data: list[OHLCVBar]) -> IndicatorResult:
        """Calculate ADX and return as IndicatorResult."""
        if not ohlcv_data:
            return IndicatorResult(
                indicator_type=IndicatorType.ADX,
                timestamp=0,
                values=IndicatorValue(single=0.0),
                params=self.params,
                error="No data",
            )

        timeperiod_value = self.params.get("timeperiod", 14)
        if not isinstance(timeperiod_value, (int, float)):
            timeperiod_value = 14
        timeperiod = int(timeperiod_value)
        if len(ohlcv_data) < timeperiod:
            return IndicatorResult(
                indicator_type=IndicatorType.ADX,
                timestamp=ohlcv_data[-1]["timestamp"],
                values=IndicatorValue(single=0.0),
                params=self.params,
                error="Insufficient data",
            )

        # Extract lists for the calculator
        highs = [bar["high"] for bar in ohlcv_data]
        lows = [bar["low"] for bar in ohlcv_data]
        closes = [bar["close"] for bar in ohlcv_data]

        # Use the calculator - returns full time series
        result = calculate_adx(highs, lows, closes, period=timeperiod)
        adx_series = result.get("adx", [])

        # Get the last value from the series
        if adx_series and len(adx_series) > 0:
            latest_adx = adx_series[-1]
        else:
            latest_adx = None

        if latest_adx is None:
            latest_adx = 0.0

        return IndicatorResult(
            indicator_type=IndicatorType.ADX,
            timestamp=ohlcv_data[-1]["timestamp"],
            values=IndicatorValue(single=latest_adx),
            params=self.params,
        )

    def _should_pass_filter(self, indicator_result: IndicatorResult) -> bool:
        """Pass filter if ADX is above minimum threshold."""
        if indicator_result.error or not hasattr(indicator_result.values, "single"):
            return False

        latest_adx = indicator_result.values.single
        min_adx = self.params.get("min_adx", 25.0)
        if not isinstance(min_adx, (int, float)):
            min_adx = 25.0

        return latest_adx >= min_adx

File: backend/src/nodes/core/market/filters/adx-filter-node.ts
// src/nodes/core/market/filters/adx-filter-node.ts
// Translated from: nodes/core/market/filters/adx_filter_node.py

import { BaseIndicatorFilter } from './base/base-indicator-filter-node';
import { IndicatorType, createIndicatorResult, createIndicatorValue } from '../../../../core/types';
import type {
  ParamMeta,
  DefaultParams,
  OHLCVBar,
  IndicatorResult,
} from '../../../../core/types';
import { calculateAdx } from '../../../../services/indicator-calculators';

/**
 * Filters assets based on ADX (Average Directional Index) values.
 */
export class ADXFilter extends BaseIndicatorFilter {
  static override defaultParams: DefaultParams = {
    min_adx: 25.0,
    timeperiod: 14,
  };

  static override paramsMeta: ParamMeta[] = [
    {
      name: 'min_adx',
      type: 'number',
      default: 25.0,
      min: 0.0,
      max: 100.0,
      step: 0.1,
    },
    { name: 'timeperiod', type: 'number', default: 14, min: 1, step: 1 },
  ];

  protected override validateIndicatorParams(): void {
    const minAdx = this.params.min_adx ?? 25.0;
    const timeperiod = this.params.timeperiod ?? 14;

    if (typeof minAdx !== 'number' || minAdx < 0) {
      throw new Error('Minimum ADX cannot be negative');
    }
    if (typeof timeperiod !== 'number' || timeperiod <= 0) {
      throw new Error('Time period must be positive');
    }
  }

  protected calculateIndicator(ohlcvData: OHLCVBar[]): IndicatorResult {
    if (!ohlcvData || ohlcvData.length === 0) {
      return createIndicatorResult(
        IndicatorType.ADX,
        0,
        createIndicatorValue(0.0),
        this.params,
        'No data'
      );
    }

    let timeperiodValue = this.params.timeperiod ?? 14;
    if (typeof timeperiodValue !== 'number') {
      timeperiodValue = 14;
    }
    const timeperiod = Math.floor(timeperiodValue);

    if (ohlcvData.length < timeperiod) {
      return createIndicatorResult(
        IndicatorType.ADX,
        ohlcvData[ohlcvData.length - 1].timestamp,
        createIndicatorValue(0.0),
        this.params,
        'Insufficient data'
      );
    }

    // Extract lists for the calculator
    const highs = ohlcvData.map((bar) => bar.high);
    const lows = ohlcvData.map((bar) => bar.low);
    const closes = ohlcvData.map((bar) => bar.close);

    // Use the calculator - returns full time series
    const result = calculateAdx(highs, lows, closes, timeperiod);
    const adxSeries = result.adx;

    // Get the last value from the series
    let latestAdx: number | null = null;
    if (adxSeries && adxSeries.length > 0) {
      latestAdx = adxSeries[adxSeries.length - 1];
    }

    if (latestAdx === null) {
      latestAdx = 0.0;
    }

    return createIndicatorResult(
      IndicatorType.ADX,
      ohlcvData[ohlcvData.length - 1].timestamp,
      createIndicatorValue(latestAdx),
      this.params
    );
  }

  protected shouldPassFilter(indicatorResult: IndicatorResult): boolean {
    if (indicatorResult.error || indicatorResult.values.single === undefined) {
      return false;
    }

    const latestAdx = indicatorResult.values.single;
    let minAdx = this.params.min_adx ?? 25.0;
    if (typeof minAdx !== 'number') {
      minAdx = 25.0;
    }

    return latestAdx >= minAdx;
  }
}
